server/src/main/java/org/opensearch/index/store/RemoteSegmentStoreDirectory.java

```diff
@@ -339,6 +362,137 @@ public IndexInput openInput(String name, IOContext context) throws IOException {
         }
     }
 
+    /**
+     * Copies a list of files from the source directory to a remote based on multi-stream upload support.
+     * If vendor plugin supports uploading multiple parts in parallel, <code>BlobContainer#writeBlobByStreams</code>
+     * will be used, else, the legacy {@link RemoteSegmentStoreDirectory#copyFrom(Directory, String, String, IOContext, boolean)}
+     * will be called.
+     *
+     * @param from          The directory for all files to be uploaded
+     * @param files         A list containing the names of all files to be uploaded
+     * @param context       IOContext to be used to open IndexInput to files during remote upload
+     * @param uploadListener An {@link UploadListener} for tracking file uploads
+     * @throws Exception When upload future creation fails or if {@link RemoteSegmentStoreDirectory#copyFrom(Directory, String, String, IOContext, boolean)}
+     *                   throws an exception
+     */
+    public boolean copyFilesFrom(Directory from, Collection<String> files, IOContext context, UploadListener uploadListener)
```

Comment:
UploadListener should be a property of this class and the remote store should be bootstrapped with it else tracking of files can get inconsistent with different listeners passed